### PR TITLE
Move userspace code to sub-package in proxy.

### PR DESCRIPTION
**Pull request by BenTheElder**
Moves the userspace code in proxy to a sub-package and adds the ProxyProvider interface.

This is in preparation for landing an implementation of https://github.com/GoogleCloudPlatform/kubernetes/issues/3760, which will mostly be in another sub-package for iptables.

### CODE REVIEW

**Comment by BenTheElder** [comment]
No functionality changes, just moves the userspace code into a sub-package and makes  ServicePortName part of the shared proxy package code. Also adds the ProxyProvider interface which is just the public methods Proxier provides (as well as the forthcoming iptables implemenation).

/cc @thockin

**Comment by BenTheElder** [comment]
This is from the "good parts" of https://github.com/GoogleCloudPlatform/kubernetes/pull/9210.
Should be pretty safe, no major changes.

File: pkg/proxy/service_port_name.go [hunk 1]
```diff
@@ -18,21 +18,10 @@ package proxy
...
-	// NextEndpoint returns the endpoint to handle a request for the given
-	// service-port and source address.
-	NextEndpoint(service ServicePortName, srcAddr net.Addr) (string, error)
-	NewService(service ServicePortName, sessionAffinityType api.ServiceAffinity, stickyMaxAgeMinutes int) error
-	CleanupStaleStickySessions(service ServicePortName)
-}
-
 // ServicePortName carries a namespace + name + portname.  This is the unique
 // identfier for a load-balanced service.
 type ServicePortName struct {
```
Comment (thockin) [inline pkg/proxy/service_port_name.go]: Is there any other remaining file this can be merged with?  Or maybe just rename to types.go

**Comment by thockin** [comment]
Just one very minor nit then LGTM

File: pkg/proxy/service_port_name.go (hunk 1 above)
Comment (BenTheElder) [inline pkg/proxy/service_port_name.go]: We probably just want types.go, this was in loadbalancer.go which is now userspace.
Do we also want to move ProxyProvider to a merged types.go?

Comment (BenTheElder) [inline pkg/proxy/service_port_name.go]: merged both to types.go, push inbound.

**Comment by thockin** [comment]
LGTM

Shippable says:

# k8s.io/kubernetes/pkg/proxy/userspace
_output/local/go/src/k8s.io/kubernetes/pkg/proxy/userspace/roundrobin_test.go:70: undefined: proxy
_output/local/go/src/k8s.io/kubernetes/pkg/proxy/userspace/roundrobin_test.go:80: undefined: proxy
_output/local/go/src/k8s.io/kubernetes/pkg/proxy/userspace/roundrobin_test.go:92: undefined: proxy
_output/local/go/src/k8s.io/kubernetes/pkg/proxy/userspace/roundrobin_test.go:130: undefined: proxy
_output/local/go/src/k8s.io/kubernetes/pkg/proxy/userspace/roundrobin_test.go:157: undefined: proxy
_output/local/go/src/k8s.io/kubernetes/pkg/proxy/userspace/roundrobin_test.go:158: undefined: proxy
_output/local/go/src/k8s.io/kubernetes/pkg/proxy/userspace/roundrobin_test.go:200: undefined: proxy
_output/local/go/src/k8s.io/kubernetes/pkg/proxy/userspace/roundrobin_test.go:201: undefined: proxy
_output/local/go/src/k8s.io/kubernetes/pkg/proxy/userspace/roundrobin_test.go:291: undefined: proxy
_output/local/go/src/k8s.io/kubernetes/pkg/proxy/userspace/roundrobin_test.go:292: undefined: proxy
_output/local/go/src/k8s.io/kubernetes/pkg/proxy/userspace/roundrobin_test.go:292: too many errors
FAIL	k8s.io/kubernetes/pkg/proxy/userspace [build failed]

**Comment by BenTheElder** [comment]
Thanks, looking into it.

**Comment by BenTheElder** [comment]
Missing the ""k8s.io/kubernetes/pkg/proxy" import for
proxy.ServicePortName, double checking that I didn't goof that anywhere
else.

**Comment by BenTheElder** [comment]
Need to get in the habit of doing the full release more often, it's just
very slow in boot2docker.

**Comment by BenTheElder** [comment]
@thockin Pushed Fix.

**Comment by thockin** [comment]
LGTM

Just waiting for green

**Comment by thockin** [comment]
once this is in, maybe we can go back to the "big" PR and rebase and squash and see what else can be peeled off.

**Comment by BenTheElder** [comment]
Sounds good. Kind of want to start rewriting a bit to simplify.

**Comment by thockin** [comment]
ACK.  And I want to help out however I can so we can get past the parts we agree on and answer the hard questions :)

**Comment by BenTheElder** [comment]
Looking forward to it :smiley:

Ahh shippable is red again, flake? I'm running the release locally, I combed over the files again manually last commit but haven't finished local release tests.

**Comment by BenTheElder** [comment]
gah, not a flake, a line has a lowercase character in that last push, pushing a fixup.

**Comment by BenTheElder** [comment]
Whoo! All green :smile: :white_check_mark:

**Comment by thockin** [comment]
LGTM

@satnam6502  is oncall - it's up to him to decide if he wants to merge on a weekend.  I think it's safe for you to rebase your main branch onto this though.

**Comment by BenTheElder** [comment]
Ok, will do. Planning to copy out the comments about redesigning the rule gen, squash, rebase.

**Comment by thockin** [comment]
let me know when done, I'll start another pass

**Comment by satnam6502** [comment]
I am happy to merge tonight -- but not tomorrow.

**Comment by thockin** [comment]
go for it, if you're here.

**Comment by satnam6502** [comment]
I assume "go for it, if you're here" is directed at @BenTheElder but do correct me if I am wrong.

**Comment by BenTheElder** [comment]
I'd say go ahead and merge if you will.
Also I think that was directed at you @satnam6502, to go for the merge. Not positive though,
@thockin?

**Comment by satnam6502** [comment]
Our e2e's are not green right now and I am trying to work out why. So, actually, it might be better to wait for green and merge on Monday.

**Comment by BenTheElder** [comment]
Ok, I've been looking through the recent commits, pretty sure it wasn't me. Haven't seen any obvious causes yet though.

**Comment by satnam6502** [comment]
I think I have found the culprit. I have swapped it out and I will swap this in -- just in time for the last e2e run of the evening.

**Comment by BenTheElder** [comment]
Thanks!

**Comment by satnam6502** [comment]
@BenTheElder I am enjoying your Rust web-socket code :-)

**Comment by BenTheElder** [comment]
Ha, thanks. I really need to update that slack package sometime

**Comment by BenTheElder** [comment]
@thockin, done rebasing / squashing, only 5 files changed now. This makes things much smaller.
Going to clean up the rule generation then link it or push it (i have it in a different local branch for now) back at https://github.com/GoogleCloudPlatform/kubernetes/pull/9210

**Comment by satnam6502** [comment]
@BenTheElder since merging this (and reverting a PR from yesterday) all the e2e runs have been green since last night so we're golden (or rather, green).

**Comment by BenTheElder** [comment]
Excellent. :)

**Pull request merged**